src/scalarstats.jl

```diff
@@ -261,36 +261,73 @@ realXcY(x::Real, y::Real) = x*y
 realXcY(x::Complex, y::Complex) = real(x)*real(y) + imag(x)*imag(y)
 
 """
-    sem(x)
-
-Return the standard error of the mean of collection `x`,
-i.e. `sqrt(var(x, corrected=true) / length(x))`.
-"""
-function sem(x)
-    y = iterate(x)
-    if y === nothing
-        T = eltype(x)
-        # Return the NaN of the type that we would get, had this collection
-        # contained any elements (this is consistent with std)
-        return oftype(sqrt((abs2(zero(T)) + abs2(zero(T)))/2), NaN)
+    sem(x; mean=nothing)
+    sem(x::AbstractArray[, weights::AbstractWeights]; mean=nothing)
+
+Return the standard error of the mean for a collection `x`. When using no weights, this is
+equal to the (sample) standard deviation divided by the sample size. If weights are used, 
+the variance of the sample mean is calculated as follows:
+
+* `AnalyticWeights`: Not implemented.
+* `FrequencyWeights`: ``\\frac{\\sum_{i=1}^n w_i (x_i - \\bar{x_i})^2}{(\\sum w_i) (\\sum w_i - 1)}``
+* `ProbabilityWeights`: ``\\frac{n}{n-1} \\frac{\\sum_{i=1}^n w_i^2 (x_i - \\bar{x_i})^2}{\\left( \\sum w_i \\right)^2}``
+
+The standard error is then the square root of the above quantities.
+
+# References
+
+Carl-Erik Särndal, Bengt Swensson, Jan Wretman (1992). Model Assisted Survey Sampling.
+New York: Springer. pp. 51-53.
+"""
+function sem(x; mean=nothing)
+    n = 0
+    if isempty(x)
+        # Return the NaN of the type that we would get for a nonempty x
+        variance = var(x; mean=mean, corrected=true)
+    elseif mean !== nothing
+        sse = real(zero(mean))
+        for element in x
+            n += 1
+            sse += abs2(element - mean)
+        end
+        variance = sse / (n - 1)
+    else
+        # Use Welford algorithm as seen in (among other places)
+        # Knuth's TAOCP, Vol 2, page 232, 3rd edition.
+        mean = zero(first(x))/1
+        sse = real(mean)
+        for element in x
+            n += 1
+            new_mean = mean + (element - mean) / n
+            sse += realXcY(element - mean, element - new_mean)
+            mean = new_mean
+        end
+        variance = sse / (n - 1)
```

Comment:
This is still not the current implementation... As @nalimilan said there's no compelling reason to deviate from the current implementation (and hence the one in Statistics). For instance, now you introduced a possible bug since the call `first(x)` may already mutate the iterable (usually it doesn't but the current code can deal with it whereas yours can't). A prominent example is e.g. `EachLine` (basically taken from the docs of `eachline`):
```julia
julia> open("my_file.txt", "w") do io
           write(io, "JuliaLang is a GitHub organization.\n It has many members.\n");
       end;

julia> for line in eachline("my_file.txt")
           print(line)
       end
JuliaLang is a GitHub organization. It has many members.

julia> lines = eachline("my_file.txt");

julia> first(lines)
"JuliaLang is a GitHub organization."

julia> for line in lines
           print(line)
       end
 It has many members.

julia> rm("my_file.txt")
```

Reply:
Thanks, fixed that. 

That's pretty horrifying behavior for `first`.